File: blaze_flutter_sdk/android/src/main/kotlin/com/blaze/flutterblazesdk/players/videos/BlazeReactVideosPlayerStyleMerger.kt
package com.blaze.flutterblazesdk.players.videos

import android.content.Context
import com.blaze.blazesdk.style.players.videos.BlazeVideosPlayerButtonsStyle
import com.blaze.blazesdk.style.players.videos.BlazeVideosPlayerCtaIconStyle
import com.blaze.blazesdk.style.players.videos.BlazeVideosPlayerCtaStyle
import com.blaze.blazesdk.style.players.videos.BlazeVideosPlayerHeadingTextStyle
import com.blaze.blazesdk.style.players.videos.BlazeVideosPlayerSeekBarStyle
import com.blaze.blazesdk.style.players.videos.BlazeVideosPlayerStyle
import com.blaze.blazesdk.style.shared.models.blazeDp
import com.blaze.flutterblazesdk.players.shared.mergeButtonThemes
import com.blaze.flutterblazesdk.players.shared.mergedWith
import com.blaze.flutterblazesdk.utils.parsing.safeParseColor
import com.blaze.flutterblazesdk.utils.parsing.toFontResId
import com.blaze.flutterblazesdk.utils.parsing.toImageResId

fun BlazeVideosPlayerStyle.mergedWith(
    customization: BlazeReactVideosPlayerStyle?,
    context: Context
): BlazeVideosPlayerStyle {
    customization ?: return this

    val merged = this

    merged.headingText = this.headingText.mergedWith(customization.headingText, context)
    merged.buttons = this.buttons.mergedWith(customization.buttons, context)
    merged.cta = this.cta.mergedWith(customization.cta, context)
    safeParseColor(customization.backgroundColor)?.let { merged.backgroundColor = it }
    merged.seekBar = this.seekBar.mergedWith(customization.seekBar, context)
    return merged
}

fun BlazeVideosPlayerHeadingTextStyle.mergedWith(
    customization: BlazeReactVideosPlayerHeadingTextStyle?,
    context: Context
): BlazeVideosPlayerHeadingTextStyle {
    customization ?: return this

    val merged = this

    merged.textSize = customization.textSize ?: this.textSize
    merged.textColor = safeParseColor(customization.textColor) ?: this.textColor
    merged.fontResId = customization.font?.toFontResId(context)
    merged.contentSource = customization.contentSource?.mapToBlazeEnumClass() ?: this.contentSource
    merged.isVisible = customization.isVisible ?: this.isVisible
    merged.maxLines = customization.numberOfLines ?: this.maxLines

    return merged
}

fun BlazeVideosPlayerButtonsStyle.mergedWith(
    customization: BlazeReactVideosPlayerButtonsStyle?,
    context: Context
): BlazeVideosPlayerButtonsStyle {
    customization ?: return this

    val merged = this
    merged.mute = merged.mute.mergeButtonThemes(customization.mute, context)
    merged.exit = merged.exit.mergeButtonThemes(customization.exit, context)
    merged.share = merged.share.mergeButtonThemes(customization.share, context)
    merged.like = merged.like.mergeButtonThemes(customization.like, context)
    merged.playPause = merged.playPause.mergeButtonThemes(customization.playPause, context)
    merged.previous = merged.previous.mergeButtonThemes(customization.previous, context)
    merged.next = merged.next.mergeButtonThemes(customization.next, context)
    return merged
}

fun BlazeVideosPlayerCtaStyle.mergedWith(
    customization: BlazeReactVideosPlayerCtaStyle?,
    context: Context
): BlazeVideosPlayerCtaStyle {
    customization ?: return this

    val merged = this
    merged.cornerRadius = customization.cornerRadius?.blazeDp ?: this.cornerRadius
    merged.textSize = customization.textSize ?: this.textSize
    merged.fontResId = customization.font?.toFontResId(context = context)
    merged.width = customization.width?.blazeDp ?: this.width
    merged.height = customization.height?.blazeDp ?: this.height
    customization.icon?.let { this.icon = this.icon.mergeWith(it, context) }
    merged.visibility = this.visibility.mergedWith(customization.ctaVisibility)
    return merged
}

fun BlazeVideosPlayerCtaStyle.BlazeCTAVisibility.mergedWith(
    customization: Map<String, Any>?
): BlazeVideosPlayerCtaStyle.BlazeCTAVisibility {
    customization ?: return this

    val type = customization["type"] as? String ?: return this
    when (type) {
        "alwaysVisible" -> return BlazeVideosPlayerCtaStyle.BlazeCTAVisibility.AlwaysVisible
        "visibleAfterOverlayHidden" -> {
            val seconds = customization["duration"] as? Double ?: return this
            return BlazeVideosPlayerCtaStyle.BlazeCTAVisibility.VisibleAfterOverlayHidden(seconds)
        }
        else -> return this
    }
}

fun BlazeVideosPlayerCtaIconStyle?.mergeWith(
    customization: BlazeReactVideosPlayerCtaIconStyle?,
    context: Context
): BlazeVideosPlayerCtaIconStyle? {
    customization ?: return this
    val imageResId = customization.iconImage?.toImageResId(context)

    val iconPositioning = customization.iconPositioning?.mapToBlazeEnumClass()

    val merged = this

    merged?.let { iconStyle ->
        merged.iconTint = safeParseColor(customization.iconTint) ?: iconStyle.iconTint
        merged.iconPositioning = iconPositioning ?: iconStyle.iconPositioning
        merged.iconResId = imageResId ?: iconStyle.iconResId
    }
        ?: run {
            imageResId ?: return null
            iconPositioning ?: return null

            BlazeVideosPlayerCtaIconStyle(
                iconResId = imageResId,
                iconPositioning = iconPositioning,
                iconTint = safeParseColor(customization.iconTint) ?: this?.iconTint
            )
        }

    return merged
}

fun BlazeVideosPlayerSeekBarStyle.mergedWith(
    customization: BlazeReactVideosPlayerSeekBarStyle?,
    context: Context
): BlazeVideosPlayerSeekBarStyle {
    customization ?: return this

    val merged = this

    merged.isVisible = customization.isVisible ?: this.isVisible
    merged.playingState = merged.playingState.mergedWith(customization.playingState, context)
    merged.pausedState = merged.pausedState.mergedWith(customization.pausedState, context)
    merged.horizontalMargin = customization.horizontalSpacing?.blazeDp ?: merged.horizontalMargin
    merged.bottomMargin = customization.bottomSpacing?.blazeDp ?: this.bottomMargin

    return merged
}

File: blaze_flutter_sdk/ios/Sources/Players/Videos/BlazeReactVideosPlayerStyleMerger.swift
//
//  BlazeReactVideosPlayerStyleMerger.swift
//  blaze_flutter_sdk
//
//

import BlazeSDK
import Foundation

extension BlazeVideosPlayerStyle {
    func mergedWith(_ customization: BlazeReactVideosPlayerStyle?) -> Self {
        guard let customization else { return self }

        var merged = self
        merged.headingText = merged.headingText.mergedWith(customization.headingText)
        merged.buttons = merged.buttons.mergedWith(customization.buttons)
        merged.cta = merged.cta.mergedWith(customization.cta)
        merged.backgroundColor =
            UIColor(hexString: customization.backgroundColor) ?? merged.backgroundColor
        merged.seekBar = merged.seekBar.mergedWith(customization.seekBar)
        return merged
    }
}

extension BlazeVideosPlayerSeekBarStyle {
    func mergedWith(_ customization: BlazeReactVideosPlayerSeekBarStyle?) -> Self {
        guard let customization else { return self }

        var merged = self
        merged.isVisible = customization.isVisible ?? merged.isVisible
        merged.playingState = merged.playingState.mergedWith(customization.playingState)
        merged.pausedState = merged.pausedState.mergedWith(customization.pausedState)
        merged.horizontalSpacing = customization.horizontalSpacing ?? merged.horizontalSpacing
        merged.bottomSpacing = customization.bottomSpacing ?? merged.bottomSpacing
        return merged
    }
}

extension BlazeVideosPlayerCtaStyle {
    func mergedWith(_ customization: BlazeReactVideosPlayerCtaStyle?) -> Self {
        guard let customization else { return self }

        var merged = self
        merged.cornerRadius = customization.cornerRadius ?? merged.cornerRadius
        merged.height = customization.height ?? merged.height
        merged.width = customization.width ?? merged.width
        merged.font = merged.font.fontWith(customization.font, textSize: customization.textSize)
        merged.icon = UIImage.imageWith(customization.icon?.iconImage)
        return merged
    }
}

extension BlazeVideosPlayerButtonsStyle {
    func mergedWith(_ customization: BlazeReactVideosPlayerButtonsStyle?) -> Self {
        guard let customization else { return self }

        var merged = self
        merged.exit = merged.exit.mergedWith(customization.exit)
        merged.like = merged.like.mergedWith(customization.like)
        merged.mute = merged.mute.mergedWith(customization.mute)
        merged.playPause = merged.playPause.mergedWith(customization.playPause)
        merged.share = merged.share.mergedWith(customization.share)
        merged.previous = merged.previous.mergedWith(customization.previous)
        merged.next = merged.next.mergedWith(customization.next)
        return merged
    }
}

extension BlazeVideosPlayerButtonStyle {
    func mergedWith(_ customization: BlazeReactPlayerButtonStyle?) -> Self {
        guard let customization else { return self }

        var merged = self
        merged.color = UIColor(hexString: customization.color) ?? merged.color
        merged.height = customization.height ?? merged.height
        merged.width = customization.width ?? merged.width
        merged.isVisible = customization.isVisible ?? merged.isVisible
        merged.isVisibleForAds = customization.isVisibleForAds ?? merged.isVisibleForAds
        merged.customImage = merged.customImage.mergedWith(customization.customImage)
        merged.contentHorizontalAlignment = .center
        merged.contentVerticalAlignment = .center

        return merged
    }
}

extension BlazeVideosPlayerHeadingTextStyle {
    func mergedWith(_ customization: BlazeReactVideosPlayerHeadingTextStyle?) -> Self {
        guard let customization else { return self }

        var merged = self

        merged.textColor = UIColor(hexString: customization.textColor) ?? merged.textColor
        merged.font = merged.font.fontWith(customization.font, textSize: customization.textSize)
        merged.isVisible = customization.isVisible ?? merged.isVisible
        merged.contentSource = merged.contentSource.mergedWith(customization.contentSource)
        merged.numberOfLines = customization.numberOfLines ?? merged.numberOfLines

        return merged
    }
}

extension BlazeVideosPlayerHeadingTextStyle.ContentSource {

    func mergedWith(_ customization: BlazeReactVideosPlayerHeadingTextStyle.ContentSource?) -> Self
    {
        guard let customization else { return self }

        switch customization {
        case .title:
            return .title
        }
    }

}

extension BlazeSDK.BlazeVideosPlayerButtonCustomImageStates? {

    func mergedWith(_ customization: BlazeReactPlayerButtonCustomImageStates?) -> Self {
        guard let customization else { return self }

        guard let defaultImage = UIImage.imageWith(customization.unselectedImage) else {
            return self
        }

        return .init(
            default: defaultImage,
            selected: UIImage.imageWith(customization.selectedImage))
    }

}
